RFC/src/RFC-0340_VNConsensusOverview.md

```diff
@@ -1,6 +1,244 @@
+view. Every view goes through three phases (`Prepare`, `PreCommit`, `Commit`) before final consensus is reached. Once a
+view reaches the `Commit` phase, it is finalised and will never be reversed.
+
+As part of their normal operation, every replica broadcasts [instructions] it receives for its contract to its peers.
+These instructions are stored in a replica's instruction mempool.
+
+When the [leader selection](#leader-selection) function designates a replica as leader for the next view, it will try
+and execute _all_ the instructions it currently has in its mempool to update the state for the next view. Following this
+it compiles a tuple of <_valid-instructions_, _rejected-instructions_, _new-state_>. This tuple represents the `CMD`
+structure described in [HotStuff].
+
+In parallel with this, the leader expects a set of `NewView` messages from the other replicas, indicating that the other
+replicas know that this replica is the leader for the next view.
+
+Once a super-majority of these messages have been received, the leader composes a proposal for the next state by adding
+a new node to the state history graph (I'm calling it a state history graph to avoid naming confusion, but it's really a
+blockchain). It composes a message containing the new proposal, and broadcasts it to the other replicas.
+
+Replicas, on receipt on the proposal, decide whether the proposal is valid, both from a protocol point of view (i.e. did
+the leader provide a well-formed proposal) as well as whether they agree on the new state (e.g. by executing the
+instructions as given and comparing the resulting state with that of the proposal). If there is agreement, they vote on
+the proposal by signing it, and send their partial signature back to the leader.
+
+When the leader has received a super-majority of votes, it sends a message back to the replicas with the (aggregated)
+set of signatures.
+
+Replicas can validate this signature and provide another partial signature indicating that they've received the first
+aggregated signature for the proposal.
+
+At this point, all replicas know that enough other replicas have received the proposal and are in agreement that it is
+valid.
+
+In Tendermint, replicas would now wait for the timeout period to make sure that the proposal wasn't going to be
+superseded before finalising the proposal. But there is an attack described in the [HotStuff] paper that could stall
+progress at this point.
+
+The HotStuff protocol prevents this by having a final round of confirmations with the leader. This prevents the stalling
+attack and _also_ lets replicas finalise the proposal _immediately_ on receipt of the final confirmation from the
+leader. This lets HotStuff proceed at "network" speed, rather than with a heartbeat dictated by the BFT synchronicity
+parameter.
+
+Although there are 4 round trips of communication between replicas and the leader, the number of messages sent are O(n).
```

Comment:
This could be trouble some if we have some branching. If there isnt any weirdness or branching thou this should be fine

Reply:
As discussed in IRC, it's required that every honest replica gets the same output given the same input state, instruction set and smart contract, so each replica follows the same code path.

This disallows things like pure random functions, for instance.

Reply:
This should not happen, but we should take care to handle this is if it does happen. 
So if we have 4 rounds happening simultaneously, with for example: one gathering instructions, one performing it, one signing, one finalising. If one of those rounds fail, the other following it needs to reset or will fail, as they were done on the assumption that the other rounds where completed.  You can reduce the risk by carefully selecting instructions.

Reply:
You have to allow for this (forks and rollbacks within the 4 round voting process)-- it's integral to the consensus mechanism. If 2/3 of nodes are non-Byzantine, then by the very nature of the protocol, there won't be any problems, irrespective of the instruction set the leader chooses.